bblocks/import_tools/imf.py

```diff
@@ -0,0 +1,175 @@
+        url (str): URL to parse
+        concat_url: base url onto which a href is attached
+
+    Returns:
+        A dictionary with dates as keys and URLs as values
+    """
+
+    try:
+        content = requests.get(url).content
+    except ConnectionError:
+        raise ConnectionError(f'Could not read page: {url}')
+
+    soup = BeautifulSoup(content, 'html.parser')
+    table = soup.find_all('table')[4]
+    links = table.find_all('a')
+
+    return {link.string: f'{concat_url}{link.get("href")}' for link in links}
+
+
+def _get_tsv_url(url: str) -> str:
+    """retrieves the URL for the SDR TSV file
+
+    Args:
+        url (str): URL to parse
+
+    Returns:
+        A link to the TSV file
+    """
+
+    try:
+        content = requests.get(url).content
+    except ConnectionError:
+        raise ConnectionError(f'could not read page: {url}')
+    soup = BeautifulSoup(content, "html.parser")
+    href = soup.find_all("a", string="TSV")[0].get("href")
+
+    return f"https://www.imf.org/external/np/fin/tad/{href}"
+
+
+def _read_sdr_tsv(url: str) -> pd.DataFrame:
+    """Reads a TSV file with SDR data from the web to a pandas dataframe
+
+    Args:
+        url (str): link to the TSV file
+
+    Returns:
+        A pandas dataframe with SDR data
+    """
+
+    df = pd.read_csv(url, delimiter="/t", engine="python").loc[3:]
+    df = df["SDR Allocations and Holdings"].str.split("\t", expand=True)
+    df.columns = ["member", "holdings", "allocations"]
+
+    return (df.melt(id_vars='member', value_vars=["holdings", "allocations"])
+            .pipe(clean_numeric_series, series_columns="value")
+            .rename(columns={"variable": "indicator"})
+            .reset_index(drop=True)
+            )
+
+
+def _check_indicators(indicators: str | list) -> list:
+    """ """
+
+    accepted_indicators = ['allocations', 'holdings']
+    if isinstance(indicators, str):
+        if indicators not in accepted_indicators:
+            raise ValueError(f"{indicators} is not a valid indicator")
+        return [indicators]
+    elif isinstance(indicators, list):
+        for indicator in indicators:
+            if indicator not in accepted_indicators:
+                raise ValueError(f"{indicator} is not a valid indicator")
+        return indicators
+    elif indicators is None:
+        return ['allocations', 'holdings']
+
+
+class SDR(ImportData):
+    """ """
+
+    indicators: list = field(default_factory=list)
+
+    def load_indicator(self, indicators: Optional | str = None) -> ImportData:
+        """ """
+
+        # make sure indicators are valid
+        indicator_list = _check_indicators(indicators)
+
+        if not os.path.exists(f"{PATHS.imported_data}/{self.file_name}") or self.update_data:
+            self.update()
+
+        df = pd.read_csv(f"{PATHS.imported_data}/{self.file_name}")
+        self.data = df.loc[df['indicator'].isin(indicator_list)].reset_index(drop=True)
+        self.indicators = indicator_list
+
```

Comment:
You would never have a list of indicators in this case as, as above, load indicator takes a string. This is where it is handy that `'indicators` is a dictionary. That way you can store dataframes with the right key